plugins/com.google.cloud.tools.eclipse.appengine.newproject.maven/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenAppEngineStandardWizardPage.java

```diff
@@ -344,34 +359,30 @@ public void modifyText(ModifyEvent event) {
   }
 
   /**
-   * Auto-fills {@link #javaPackageField} as Group ID when
-   * 1) {@link #javaPackageField} is empty; or
-   * 2) the field matches previous auto-fill before ID modification.
+   * Auto-fills {@link #javaPackageField} from the Group ID if the user has not explicitly chosen a
+   * package name.
    */
-  private final class AutoPackageNameSetterOnGroupIdChange implements VerifyListener {
-
-    private String previousSuggestion = "";
+  private final class AutoPackageNameSetterOnGroupIdChange implements ModifyListener {
 
     @Override
-    public void verifyText(VerifyEvent event) {
-      String groupId = groupIdField.getText();
-      // Below explains how to get text after modification:
-      // http://stackoverflow.com/questions/32872249/get-text-of-swt-text-component-before-modification
-      String newGroupId =
-          groupId.substring(0, event.start) + event.text + groupId.substring(event.end);
-
+    public void modifyText(ModifyEvent event) {
       // getGroupId() trims whitespace, so we do the same to sync with the dialog validation error.
-      if (MavenCoordinatesValidator.validateGroupId(newGroupId.trim())) {
-        String newSuggestion = suggestPackageName(newGroupId);
+      String groupId = groupIdField.getText().trim();
+
+      if (MavenCoordinatesValidator.validateGroupId(groupId)) {
+        String newSuggestion = suggestPackageName(groupId);
         updatePackageField(newSuggestion);
-        previousSuggestion = newSuggestion;
+      } else if (groupId.isEmpty()) {
+        updatePackageField("");
       }
     }
+  }
 
-    private void updatePackageField(String newSuggestion) {
-      if (getPackageName().isEmpty() || getPackageName().equals(previousSuggestion)) {
-        javaPackageField.setText(newSuggestion);
-      }
+  private void updatePackageField(String newSuggestion) {
+    if (autoGeneratePackageName) {
```

Comment:
Why don't we check this even before we generate the suggestion in `modifyText()`?